src/xform_to_datalog/arcs_manifest_tree/particle_spec.h

```diff
@@ -0,0 +1,66 @@
+#ifndef SRC_XFORM_TO_DATALOG_ARCS_MANIFEST_TREE_PARTICLE_SPEC_H_
+#define SRC_XFORM_TO_DATALOG_ARCS_MANIFEST_TREE_PARTICLE_SPEC_H_
+
+#include <string>
+
+#include "absl/container/flat_hash_map.h"
+#include "src/common/logging/logging.h"
+#include "src/ir/edge.h"
+#include "src/ir/tag_check.h"
+#include "src/ir/tag_claim.h"
+#include "src/xform_to_datalog/arcs_manifest_tree/handle_connection_spec.h"
+#include "third_party/arcs/proto/manifest.pb.h"
+
+namespace raksha::xform_to_datalog::arcs_manifest_tree {
+
+// A class representing a ParticleSpec. This contains the relevant
+// information for the dataflow graph: checks, claims, and edges within the
+// ParticleSpec. All such information uses this ParticleSpec as the root.
+// This means that it cannot be printed to Datalog facts until these items
+// are instantiated with a fully-instantiated root.
+class ParticleSpec {
+ public:
+
+  // Create the ParticleSpec from a ParticleSpecProto.
+  static ParticleSpec CreateFromProto(
+      const arcs::ParticleSpecProto &particle_spec_proto);
+
+  const std::string &name() const { return name_; }
+
+  const std::vector<raksha::ir::TagCheck> &checks() const { return checks_; }
+
+  const std::vector<raksha::ir::TagClaim> &claims() const { return claims_; }
+
+  const std::vector<raksha::ir::Edge> &edges() const { return edges_; }
+
+ private:
+  ParticleSpec(
+      std::string name,
+      std::vector<raksha::ir::TagCheck> checks,
+      std::vector<raksha::ir::TagClaim> claims,
+      std::vector<raksha::ir::Edge> edges)
+      : name_(std::move(name)), checks_(std::move(checks)),
+        claims_(std::move(claims)), edges_(std::move(edges)) { }
+
+  // Generate the edges between HandleConnectionSpecs within this ParticleSpec.
+  static std::vector<raksha::ir::Edge> GenerateEdges(
+      std::string particle_spec_name,
+      std::vector<HandleConnectionSpec> connection_specs);
+
+  // The name of this ParticleSpec.
```

Comment:
I think you might also want to capture the inputs and outputs of the particle spec. OK to defer it to when the need arises.

Reply:
Sounds good. One of the reasons I'm deferring it is because, as written, `HandleConnectionSpec`s contain a `std::unique_ptr<Type>`, which makes them a bit harder to work with. That said, you can imagine a refactoring in which `HandleConnectionSpec` just performs the generation of its `AccessPath` set eagerly, allowing for holding the `unique_ptr` for an even briefer period of time.